mDCV bug: the FP API set incorrect values

jbowler opened the pull request:
BUG: png_set_mDCV chromaticities set incorrectly

png_set_mDCV was dividing the chromaticities by two.

Signed-off-by: John Bowler <jbowler@acm.org>

jbowler committed:
mDCV bug: the FP API set incorrect values

png_set_mDCV was dividing the chromaticities by two.

Signed-off-by: John Bowler <jbowler@acm.org>

ctruta reviewed (changes requested):
Re-reviewing...

Umm, actually, this is only half-fixed. (Considering the nature of this defect, intending a pun is rather appropriate.)

Only the setter has been fixed, but the getter needs fixing, too. (I'll do it. This calls for an immediate fix, as well as an immediate release of libpng v1.6.48.)

ctruta reviewed (approved):
Uh...
Re-re-re-...-reviewing: the fix by @jbowler is correct, only the setter was broken, and the differences in how `png_get_mDCv*` are implemented vs. how `png_set_mDCv*` are implemented makes the code extra-confusing.

jbowler commented:
I checked both.  The explanation is in the #699 where I explain how the problem arose.

To elaborate on the principle.  The "fixed" APIs are normally the principle ones because originally relying on hardware with adequate FP support produced problems (unsupported APIs) on lower end systems.  The parameters being set are also invariably integral because PNG does not use a native FP (like IEEE 488 stuff); going via FP introduces the complexity of checking for rounding errors and requires using double because float does not have a full 32 bits of precision (not that this matters much in practice).

Consequently at present almost all the checks are implemented in the fixed APIs to ensure code commonality and, more important, bug commonality if the checks are wrong.

So I changed png_set_mDCV to call png_set_mDCV_fixed without undoing the divide that was there originally.  Oops.

It's a PITA that the W3C reinvented the cHRM chunk with a different format and a different set of error/range conditions.  Totally pointless.  It would have been much, much, better to fix the original problem with cHRM by saying that cHRM **is** the mastering display colour volume (it is if there is no mDCV) and adding a proper CIEXYZTRIPLE (in Windows terms) chunk for the encoding colour space (which defaults to the horrible matrix invert back calculation from cHRM at present).

ctruta commented:
@jbowler wrote:

> I checked both. The explanation is in the #699 where I explain how the problem arose.

I think you meant #669. But otherwise, I hear you.

> So I changed png_set_mDCV to call png_set_mDCV_fixed without undoing the divide that was there originally. Oops.

This is an rather solid indication that the pngtest program needs to check both the fixed-point API and the floating-point API. It should not be unnecessarily restricted to check just one, and worse, to be restricted at compile time. At the very least, there should be run-time option, "check the fixed-point API" vs "check the floating-point API", and then we should run pngtest in our test suite with this flag on and off. FYI I already started implementing that. Not for v1.6.48 (soon to be released) but maybe for later, and definitely for v1.8.0.

jbowler commented:
> This is an rather solid indication that the pngtest program needs to check both the fixed-point API and the floating-point API. It should not be unnecessarily restricted to check just one, and worse, to be restricted at compile time.

It's the same as `contrib/conftest/{read,write}-full.dfa` - I added those because even though there were tests in contrib/conftest for both FP and fixed APIs there were confined with "likely" minimal configurations so skipped many of the chunks.

See `contrib/conftest/float-fixed.dfa` and `contrib/conftest/fixed.dfa` but those don't take into account the fact that `pngtest` **defaults** to fixed point (because of the way the #if/#else/#endif test is written.)  At least I think that's the problem.  See line 1119 of pngtest.c.

I think a minimalist fix is to change 1119 to `#ifndef PNG_FLOATING_POINT_SUPPORTED` on the basis that one or the other is required.  Another approach is `contrib/conftest/float.dfa`:

`option FIXED_POINT off`

NOTE: I originally said "FIXED_ARITHMETIC" but there is no such option, its:

FIXED_POINT - fixed point APIs
FLOATING_POINT - floating point APIs
FLOATING _ARITHMETIC - allow internal use of FP code

Hum, won't take long to test :-)  (EDIT: `contrib/conftest/float.dfa` doesn't seem to fail as required; pngtest passes...)

jbowler commented:
Both checks work as expected with configure - my problem before was not doing a clean build.  I.e. without the png_set_mDCV fix an unconfigured build fails the pngtest check and configured builds fail as expected according to whether the floating point APIs are used (fail) or not (pass).

cmake passes but produces a bogus pngtest; i.e. cmake apparently isn't doing the binary compare because when I run the generated pngtest directly it produces a changed pngtest.out but exits with a 0 (success) exit code.   My build script also doesn't manage to do the minconfigs for cmake.  I suspect DFA_XTRA isn't working or maybe I don't understand cmake; I'm setting DFA_XTRA in the environment.

Both the cmake problems are not new.   I PRed #675 for your perusal.  It's not as good as testing both APIs at once but the "conftest" is useful on its own because it causes all the test programs (not just pngtest) to use the FP APIs (I think they default to that anyway but I haven't checked).
